compiler/rustc_codegen_llvm/src/builder.rs

```diff
@@ -24,17 +25,276 @@ use rustc_span::Span;
+    // entry:
+    //   %0 = tail call double (...) @__enzyme_autodiff_square(double (double)* nonnull @square, double %x)
+    //   ret double %0
+    // }
+    // ```
+    unsafe {
+        // On LLVM-IR, we can luckily declare __enzyme_ functions without specifying the input
+        // arguments. We do however need to declare them with their correct return type.
+        // We already figured the correct return type out in our frontend, when generating the outer_fn,
+        // so we can now just go ahead and use that. FIXME(ZuseZ4): This doesn't handle sret yet.
+        let fn_ty = llvm::LLVMGlobalGetValueType(outer_fn);
+        let ret_ty = llvm::LLVMGetReturnType(fn_ty);
+
+        // LLVM can figure out the input types on it's own, so we take a shortcut here.
+        let enzyme_ty = llvm::LLVMFunctionType(ret_ty, ptr::null(), 0, True);
+        let ad_fn = llvm::LLVMRustGetOrInsertFunction(
+            llmod,
+            ad_name.as_ptr() as *const c_char,
+            ad_name.len().try_into().unwrap(),
+            enzyme_ty,
+        );
+        // Otherwise LLVM might inline our temporary code before the enzyme pass has a chance to
+        // do it's work.
+        let attr = llvm::AttributeKind::NoInline.create_attr(llcx);
+        attributes::apply_to_llfn(ad_fn, Function, &[attr]);
+
+        // first, remove all calls from fnc
+        let entry = llvm::LLVMGetFirstBasicBlock(outer_fn);
+        let br = llvm::LLVMRustGetTerminator(entry);
+        llvm::LLVMRustEraseInstFromParent(br);
+
+        let builder = llvm::LLVMCreateBuilderInContext(llcx);
+        let last_inst = llvm::LLVMRustGetLastInstruction(entry).unwrap();
+        llvm::LLVMPositionBuilderAtEnd(builder, entry);
+
+        let num_args = llvm::LLVMCountParams(&fn_to_diff);
+        let mut args = Vec::with_capacity(num_args as usize + 1);
+        args.push(fn_to_diff);
+
+        let enzyme_const = llvm::create_md_string(llcx, "enzyme_const");
+        let enzyme_out = llvm::create_md_string(llcx, "enzyme_out");
+        let enzyme_dup = llvm::create_md_string(llcx, "enzyme_dup");
+        let enzyme_dupnoneed = llvm::create_md_string(llcx, "enzyme_dupnoneed");
+        let enzyme_primal_ret = llvm::create_md_string(llcx, "enzyme_primal_return");
+
+        match output {
+            DiffActivity::Dual => {
+                args.push(llvm::LLVMMetadataAsValue(llcx, enzyme_primal_ret));
+            }
+            DiffActivity::Active => {
+                args.push(llvm::LLVMMetadataAsValue(llcx, enzyme_primal_ret));
+            }
+            _ => {}
+        }
+
+        trace!("matching autodiff arguments");
+        // We now handle the issue that Rust level arguments not always match the llvm-ir level
+        // arguments. A slice, `&[f32]`, for example, is represented as a pointer and a length on
+        // llvm-ir level. The number of activities matches the number of Rust level arguments, so we
+        // need to match those.
```

Comment:
This index math is a bit unusual. I think I'd prefer using iterators (peekable in the `input` case), and just stepping them manually.

Reply:
I got this code wrong a few times during re-implementing / refactorings, would you be fine with leaving it for the second PR, so I can have tests to cover the changes?

Reply:
Makes sense. Add a FIXME?